contracts/utils/SignatureValidator.sol

```diff
@@ -0,0 +1,149 @@
+   * @param _data           Full EIP-712 data structure that was hashed and signed
+   * @param _sig            Proof that the hash has been signed by signer.
+   *      For non wallet signatures, _sig is expected to be an array tightly encoded as
+   *      (bytes32 r, bytes32 s, uint8 v, uint256 nonce, SignatureType sigType)
+   * @return isValid True if the address recovered from the provided signature matches the input signer address.
+   */
+  function isValidSignature(
+    address _signerAddress,
+    bytes32 _hash,
+    bytes memory _data,
+    bytes memory _sig
+  )
+    public
+    view
+    returns (bool isValid)
+  {
+    require(
+      _sig.length > 0,
+      "SignatureValidator#isValidSignature: LENGTH_GREATER_THAN_0_REQUIRED"
+    );
+
+    require(
+      _signerAddress != address(0x0),
+      "SignatureValidator#isValidSignature: INVALID_SIGNER"
+    );
+
+    // Pop last byte off of signature byte array.
+    uint8 signatureTypeRaw = uint8(_sig.popLastByte());
+
+    // Ensure signature is supported
+    require(
+      signatureTypeRaw < uint8(SignatureType.NSignatureTypes),
+      "SignatureValidator#isValidSignature: UNSUPPORTED_SIGNATURE"
+    );
+
+    // Extract signature type
+    SignatureType signatureType = SignatureType(signatureTypeRaw);
+
+    // Variables are not scoped in Solidity.
+    uint8 v;
+    bytes32 r;
+    bytes32 s;
+    address recovered;
+
+    // Always illegal signature.
+    // This is always an implicit option since a signer can create a
+    // signature array with invalid type or length. We may as well make
+    // it an explicit option. This aids testing and analysis. It is
+    // also the initialization value for the enum type.
+    if (signatureType == SignatureType.Illegal) {
+      revert("SignatureValidator#isValidSignature: ILLEGAL_SIGNATURE");
+
+
+    // Signature using EIP712
+    } else if (signatureType == SignatureType.EIP712) {
+      require(
+        _sig.length == 97,
+        "SignatureValidator#isValidSignature: LENGTH_97_REQUIRED"
+      );
+      r = _sig.readBytes32(0);
+      s = _sig.readBytes32(32);
+      v = uint8(_sig[64]);
+      recovered = ecrecover(_hash, v, r, s);
+      isValid = _signerAddress == recovered;
+      return isValid;
+
+
+    // Signed using web3.eth_sign() or Ethers wallet.signMessage()
+    } else if (signatureType == SignatureType.EthSign) {
+      require(
+        _sig.length == 97,
+        "SignatureValidator#isValidSignature: LENGTH_97_REQUIRED"
+      );
+      r = _sig.readBytes32(0);
+      s = _sig.readBytes32(32);
+      v = uint8(_sig[64]);
+      recovered = ecrecover(
+        keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", _hash)),
+        v,
+        r,
+        s
+      );
+      isValid = _signerAddress == recovered;
+      return isValid;
+
+
+    // Signature verified by wallet contract with data validation.
+    } else if (signatureType == SignatureType.WalletBytes) {
```

Comment:
We should remove this option, no-one uses it. Would allow us to not pass `_data` field as well.